session_input_callback repeating a history item re-persists it as new session input

### Please read this first

- **Have you read the docs?** Yes — [Sessions](https://openai.github.io/openai-agents-python/sessions/), [Running agents](https://openai.github.io/openai-agents-python/running_agents/), and `.agents/references/session-persistence.md`.
- **Have you searched for related issues?** Yes. Searched open/closed issues and PRs for `session_input_callback duplicate`, `session_input_callback history re-appended`, `session history duplicated callback`, `prepare_input_with_session`, `session callback persists history`, `session grows duplicated history`, and `history appended twice session`. Nearest prior work is #3852 (session loses a user turn when a retry rewinds it) and #3827 (compaction with a limited session); neither concerns callback-emitted duplicates.

### Affected component

Client-managed session persistence — `src/agents/run_internal/session_persistence.py::prepare_input_with_session`, used by `RunConfig.session_input_callback`.

### Describe the bug

If a `session_input_callback` emits an existing history item **more than once**, every extra occurrence is classified as new-turn input and written back into the session store. The stored conversation is corrupted, and because the re-persisted copy becomes part of the next turn's history, the corruption compounds on every subsequent turn.

`prepare_input_with_session()` classifies each item returned by the callback by first consuming an object-identity reference from `history_refs`/`new_refs`, then falling back to content-frequency maps. `_consume_reference()` pops the matched object out of the reference list, so the *same* history object appearing a second time no longer matches by identity; the identity match has also already decremented the content-frequency budget for that key, so the second occurrence falls through to `appended.append(item)` and is persisted as if it were fresh user input.

### Debug information

- Agents SDK version: `main` @ `c3f1781d56e8f1249a01674f18ba1f3e44a16dce` (latest release tag `v0.19.1`)
- Python version: 3.13.5
- OS: macOS 15.7.3 (Darwin 24.6.0)
- No OpenAI API key, no network access, and no paid model call are needed — the repro uses `tests/fake_model.py::FakeModel` and a temporary `SQLiteSession`.

### Repro steps

Save as `tests/test_repro_session_callback.py` and run `uv run pytest tests/test_repro_session_callback.py -q`:

```python
from __future__ import annotations

import tempfile
from pathlib import Path

import pytest

from agents import Agent, RunConfig, Runner, SQLiteSession, TResponseInputItem

from .fake_model import FakeModel
from .test_responses import get_text_message


def repeat_first(
    history: list[TResponseInputItem], new: list[TResponseInputItem]
) -> list[TResponseInputItem]:
    """Re-emphasize the original request at the end of the model context."""
    if not history:
        return new
    return history + [history[0]] + new


@pytest.mark.asyncio
async def test_repeated_history_item_is_not_re_persisted() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        session = SQLiteSession("s", Path(tmp) / "s.db")
        model = FakeModel()
        agent = Agent(name="A", model=model)
        config = RunConfig(session_input_callback=repeat_first)

        for turn in range(4):
            model.set_next_output([get_text_message(f"a{turn}")])
            await Runner.run(agent, input=f"u{turn}", session=session, run_config=config)

        stored = await session.get_items()
        user_inputs = [item.get("content") for item in stored if item.get("role") == "user"]
        # Each turn contributes exactly one user message; "u0" must not be stored again.
        assert user_inputs == ["u0", "u1", "u2", "u3"], user_inputs
        assert len(stored) == 8, len(stored)
```

### Actual behavior

Fails deterministically (reproduced 3/3 runs):

```
FAILED tests/test_repro_session_callback.py::test_repeated_history_item_is_not_re_persisted
E       AssertionError: ['u0', 'u0', 'u1', 'u0', 'u2', 'u0', ...]
E       assert ['u0', 'u0', ...2', 'u0', ...] == ['u0', 'u1', 'u2', 'u3']
```

After 4 turns the session holds 11 items instead of 8:

```
['u:u0', 'a:a0', 'u:u0', 'u:u1', 'a:a1', 'u:u0', 'u:u2', 'a:a2', 'u:u0', 'u:u3', 'a:a3']
```

The `u0` copies at indices 2, 5 and 8 were never new input — they are the same `history[0]` object the callback re-emitted. The identical behavior occurs with `Runner.run_streamed()`.

### Expected behavior

An item the callback took from the `history` argument must never be persisted as new-turn input, no matter how many times the callback emits it. Only genuinely new items belong in the append batch.

This is stated in `.agents/references/session-persistence.md`:

> Existing history must not be re-appended as new input, even when `session_input_callback` deep-copies, reorders, filters, **duplicates**, or reconstructs items.

and in the `prepare_input_with_session()` docstring itself:

> The callback may reorder, drop, or **duplicate** items. […] so retries and custom merge strategies do not accidentally re-persist old history as fresh input.

Reordering, deep-copying, filtering, and reconstructing all behave correctly today; only repetition of a history item is mishandled.

### Root-cause hypothesis (hypothesis, not verified as the maintainers' intent)

`_consume_reference()` treats history object identity as a one-shot resource: it pops the matched object from the reference list. Object identity with an item in `history_for_callback` is, however, definitive proof that the item came from history and stays true for every occurrence. Because the identity match also decrements the content-frequency budget for that key, no fallback covers the repeat, and the classifier's final `else` branch persists it.

### Proposed focused scope

Treat a history object-identity match as a durable fact rather than a consumable reference, so repeated occurrences of the same history object are pruned from the persistence batch while remaining in the model input. The content-frequency fallback for *reconstructed* history items should stay as-is, since it is what keeps a genuinely new item that coincidentally serializes identically to a history item from being dropped. Regression tests would cover repeat, reorder, deep-copy, filter, and reconstruct callbacks in both streaming and non-streaming modes, plus the case where new input duplicates history content.

### User impact

Applications whose `session_input_callback` re-injects an earlier message — a common "restate the original goal at the end of the context" or RAG-style re-injection pattern — silently and permanently corrupt their stored conversation history, and the duplicate count grows every turn. The damage persists in the session store after the process exits.

I'd like to work on this and open a PR if that's welcome.

--- a/src/agents/run_internal/session_persistence.py
+++ b/src/agents/run_internal/session_persistence.py
@@ -217,6 +217,10 @@ async def prepare_input_with_session(
             )
         history_for_callback = copy.deepcopy(converted_history)
         new_items_for_callback = copy.deepcopy(new_input_list)
+        # Keep the original history objects alive so their identities remain valid even if the
+        # callback removes them from the list it receives.
+        original_history_objects = list(history_for_callback)
+        original_history_object_ids = {id(item) for item in original_history_objects}
         combined = session_input_callback(history_for_callback, new_items_for_callback)
         if inspect.isawaitable(combined):
             combined = await combined
@@ -246,10 +250,16 @@ async def prepare_input_with_session(
             new_key = _session_item_key(item)
             if _consume_reference(new_refs, new_key, item):
                 new_counts[new_key] = max(new_counts.get(new_key, 0) - 1, 0)
-                appended.append(item)
+                if id(item) in original_history_object_ids:
+                    prune_history_indexes.add(combined_index)
+                else:
+                    appended.append(item)
                 continue
             if _consume_reference(history_refs, history_key, item):
                 history_counts[history_key] = max(history_counts.get(history_key, 0) - 1, 0)
+                prune_history_indexes.add(combined_index)
+                continue
+            if id(item) in original_history_object_ids:
                 prune_history_indexes.add(combined_index)
                 continue
             if history_counts.get(history_key, 0) > 0:

--- a/tests/memory/test_session.py
+++ b/tests/memory/test_session.py
@@ -569,6 +569,43 @@ async def test_session_callback_prepared_input(runner_method):
             session.close()
 
 
+@pytest.mark.parametrize("runner_method", ["run", "run_sync", "run_streamed"])
+@pytest.mark.asyncio
+async def test_session_callback_repeating_history_does_not_grow_session(runner_method):
+    with tempfile.TemporaryDirectory() as temp_dir:
+        db_path = Path(temp_dir) / "test_memory.db"
+        model = FakeModel()
+        agent = Agent(name="test", model=model)
+        session = SQLiteSession("session_repeat", db_path)
+
+        def repeat_first(history, new_input):
+            if not history:
+                return new_input
+            return history + [history[0]] + new_input
+
+        try:
+            for turn in range(3):
+                model.set_next_output([get_text_message(f"assistant {turn}")])
+                await run_agent_async(
+                    runner_method,
+                    agent,
+                    f"user {turn}",
+                    session=session,
+                    run_config=RunConfig(session_input_callback=repeat_first),
+                )
+
+            stored = await session.get_items()
+            user_messages = [item for item in stored if item.get("role") == "user"]
+            assert [item.get("content") for item in user_messages] == [
+                "user 0",
+                "user 1",
+                "user 2",
+            ]
+            assert len(stored) == 6
+        finally:
+            session.close()
+
+
 @pytest.mark.asyncio
 async def test_sqlite_session_unicode_content():
     """Test that session correctly stores and retrieves unicode/non-ASCII content."""

--- a/tests/test_agent_runner.py
+++ b/tests/test_agent_runner.py
@@ -2677,6 +2677,110 @@ async def test_prepare_input_with_session_matches_copied_items_by_content() -> N
 
 
 @pytest.mark.asyncio
+async def test_prepare_input_with_session_repeated_history_keeps_equal_new_item() -> None:
+    history_item = cast(TResponseInputItem, {"role": "user", "content": "same"})
+    session = SimpleListSession(history=[history_item])
+
+    def callback(
+        history: list[TResponseInputItem], new_input: list[TResponseInputItem]
+    ) -> list[TResponseInputItem]:
+        return [history[0], history[0], new_input[0]]
+
+    prepared, session_items = await prepare_input_with_session("same", session, callback)
+
+    assert [cast(dict[str, Any], item).get("content") for item in prepared] == [
+        "same",
+        "same",
+        "same",
+    ]
+    assert [cast(dict[str, Any], item).get("content") for item in session_items] == ["same"]
+
+
+@pytest.mark.asyncio
+async def test_prepare_input_with_session_async_callback_moves_repeated_history_item() -> None:
+    history_item = cast(TResponseInputItem, {"role": "user", "content": "history"})
+    session = SimpleListSession(history=[history_item])
+
+    async def callback(
+        history: list[TResponseInputItem], new_input: list[TResponseInputItem]
+    ) -> list[TResponseInputItem]:
+        await asyncio.sleep(0)
+        moved = history.pop(0)
+        return [moved, new_input[0], moved]
+
+    prepared, session_items = await prepare_input_with_session("new", session, callback)
+
+    assert [cast(dict[str, Any], item).get("content") for item in prepared] == [
+        "history",
+        "new",
+        "history",
+    ]
+    assert [cast(dict[str, Any], item).get("content") for item in session_items] == ["new"]
+
+
+@pytest.mark.asyncio
+async def test_prepare_input_with_session_history_moved_to_new_input_stays_history() -> None:
+    history_item = cast(TResponseInputItem, {"role": "user", "content": "history"})
+    session = SimpleListSession(history=[history_item])
+
+    def callback(
+        history: list[TResponseInputItem], new_input: list[TResponseInputItem]
+    ) -> list[TResponseInputItem]:
+        moved = history.pop(0)
+        new_input.insert(0, moved)
+        return new_input + [moved]
+
+    prepared, session_items = await prepare_input_with_session("new", session, callback)
+
+    assert [cast(dict[str, Any], item).get("content") for item in prepared] == [
+        "history",
+        "new",
+        "history",
+    ]
+    assert [cast(dict[str, Any], item).get("content") for item in session_items] == ["new"]
+
+
+@pytest.mark.asyncio
+async def test_prepare_input_with_session_callback_replaces_history_item() -> None:
+    history_item = cast(TResponseInputItem, {"role": "user", "content": "history"})
+    replacement = cast(TResponseInputItem, {"role": "user", "content": "summary"})
+    session = SimpleListSession(history=[history_item])
+
+    def callback(
+        history: list[TResponseInputItem], new_input: list[TResponseInputItem]
+    ) -> list[TResponseInputItem]:
+        history[0] = replacement
+        return history + new_input
+
+    prepared, session_items = await prepare_input_with_session("new", session, callback)
+
+    assert [cast(dict[str, Any], item).get("content") for item in prepared] == [
+        "summary",
+        "new",
+    ]
+    assert [cast(dict[str, Any], item).get("content") for item in session_items] == ["new"]
+
+
+@pytest.mark.asyncio
+async def test_prepare_input_with_session_extra_reconstructed_history_item_stays_new() -> None:
+    history_item = cast(TResponseInputItem, {"role": "user", "content": "history"})
+    session = SimpleListSession(history=[history_item])
+
+    def callback(
+        history: list[TResponseInputItem], new_input: list[TResponseInputItem]
+    ) -> list[TResponseInputItem]:
+        rebuilt = cast(TResponseInputItem, dict(cast(dict[str, Any], history[0])))
+        return [history[0], rebuilt, new_input[0]]
+
+    _, session_items = await prepare_input_with_session("new", session, callback)
+
+    assert [cast(dict[str, Any], item).get("content") for item in session_items] == [
+        "history",
+        "new",
+    ]
+
+
+@pytest.mark.asyncio
 async def test_prepare_input_with_openai_conversation_strips_assistant_history_ids() -> None:
     class DummyOpenAIConversationsSession(OpenAIConversationsSession):
         def __init__(self, history: list[TResponseInputItem]) -> None:

